### ISSUE TITLE
Force delete pods on timeout in WaitForControlledPodsRunning

### ISSUE THREAD

**Issue by wojtek-t** [kubernetes/perf-tests#1319]
If the node is broken (e.g. kubelet can't connect to apiserver or the node itself is hard-down or sth), controllers are not able to delete the pods (because the deletion is happening as a result of kubelet confirming that).

In order to prevent this from completely failing our tests, when WaitForControlledPodsRunning on deletion is failing with timeout, we should fail the test as we do now, but also "force delete the pods we're still waiting for".

**Comment by ZhiFeng1993** [comment]
Hi. I'm interested in working on this. @wojtek-t  do you mind adding more details? Like which test this is? Thanks!

**Comment by vamossagar12** [comment]
@wojtek-t i can take this one up. 
/assign

**Comment by vamossagar12** [comment]
@wojtek-t , i took one pass at it. As far as I understood, this is the point in gather, where we can figure out if an action timed out:
https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go#L245

Is this the place where you are expecting the logic to force delete the pods? Also, in the description, you have mentioned that 

**when WaitForControlledPodsRunning on deletion is failing with timeout**. 1 thing which I didn't understand is, how can we know that the timeout is wrt deletion or any other action?

**Comment by wojtek-t** [comment]
1) yes - that sounds reasonable

2) it seems this action is not there at this point; but when you create a checker:
https://github.com/kubernetes/perf-tests/blob/4458e55128544d9bcfa87bcb23c3019138328de9/clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go#L353
You have all the necessary information (isObjDeleted) - so it seems it's only a matter of setting a status differently maybe based on that (or some other field, probably might be better to expose - we do the casting anywway here: https://github.com/kubernetes/perf-tests/blob/4458e55128544d9bcfa87bcb23c3019138328de9/clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go#L238 )

3) you don't need kubectl - kubectl is just calling the API; you should use the API (whatever it is doing underneath - you can probably easly check it by calling kubectl delete --force and setting high-enough verbosity (--v=8) - then it prints all requests it send).

**Comment by vamossagar12** [comment]
Thanks @wojtek-t , For point 3, i was able to find the API end point being invoked:
```
I1010 15:56:53.909567   33629 request.go:968] Request Body: {"gracePeriodSeconds":0,"propagationPolicy":"Background"}
I1010 15:56:53.909620   33629 round_trippers.go:420] DELETE https://kubernetes.docker.internal:6443/api/v1/namespaces/default/pods/ssh
I1010 15:56:53.909628   33629 round_trippers.go:427] Request Headers:
I1010 15:56:53.909632   33629 round_trippers.go:431]     Accept: application/json
I1010 15:56:53.909636   33629 round_trippers.go:431]     Content-Type: application/json
I1010 15:56:53.909639   33629 round_trippers.go:431]     User-Agent: kubectl/v1.16.2 (darwin/amd64) kubernetes/c97fe50
```

This can be invoked from the coreV1 rest client end point. Only doubt i have is, how to fetch the namespace of the pods.

**Comment by wojtek-t** [comment]
I don't understand - you will need to pass --namespace too. So you would need the same there.

**Comment by vamossagar12** [comment]
What I meant was, how do I get the namespace of the pods from the measurement code?

**Comment by vamossagar12** [comment]
Thanks @wojtek-t . Created this PR: https://github.com/kubernetes/perf-tests/pull/1528

**Comment by wojtek-t** [comment]
Fixed in #1528
